File: app/modules/omarchy-link/android/src/main/java/expo/modules/omarchylink/Announce.kt
package expo.modules.omarchylink

import android.content.Context
import android.net.wifi.WifiManager
import android.os.PowerManager
import org.json.JSONObject
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetSocketAddress

/**
 * Listening for the desktop saying it is back.
 *
 * The phone's reconnect ladder tops out at two minutes and then knocks on that
 * rung forever — but the ladder is `setTimeout`s inside a headless task, and
 * `LinkService` deliberately holds no wake lock, so on a sleeping handset the
 * next knock happens whenever something else wakes the phone. Turning the
 * desktop on in the morning is not one of those somethings, which is the whole
 * of the bug: the desktop has been up for a quarter of an hour and the phone
 * has not noticed.
 *
 * The desktop cannot ring this phone. It holds an id, a name and a token for
 * it and no address of any kind — by design; the phone is always the side that
 * dials. What it can do is put one short burst of UDP on the subnet's
 * broadcast address when its daemon starts, which is what this reads.
 *
 * Three things make that work on a phone that is asleep, and all three are
 * needed:
 *
 *   - **Somewhere to receive it.** A blocking `receive()` on a thread of the
 *     foreground service. The service is what keeps the process alive; the
 *     thread costs nothing while no packet is arriving.
 *   - **A radio that will deliver it.** Wi-Fi filters out frames not addressed
 *     to this device when it is in power save, and a subnet broadcast is
 *     exactly such a frame. A `MulticastLock` is what turns that filter off;
 *     it is famously named for multicast and has always covered broadcast too.
 *   - **A CPU awake long enough to act.** The packet arriving wakes the
 *     processor for as long as it takes to hand it over and no longer, and
 *     "act" here means a TCP connect and a TLS handshake. So a wake lock is
 *     taken — but a timed one, held for seconds around one announcement, which
 *     is a different thing from the untimed lock `LinkService` refuses to hold
 *     for the life of the link.
 *
 * The packet is a hint and never a credential. Everything in it is what the
 * daemon already gives away to an unauthenticated `GET /api/info`, the key in
 * it is used as a filter and not as proof, and all this asks JavaScript for is
 * a redial of the desktop it already pinned. Anyone on the subnet can forge
 * one; the worst they achieve is the phone dialling the desktop it wanted.
 */
object Announce {
  /** Mirrors `ANNOUNCE_PORT` in the daemon's `lib/announce`. */
  const val PORT = 8766

  private const val APP = "omarchy-connect"
  private const val KIND = "desktop-up"

  /** Nothing this app is interested in is larger; a truncated read is a drop. */
  private const val MAX = 2048

  /**
   * The desktop sends three packets so that one lost frame does not lose the
   * announcement. This is the other end of that: one event out of a burst.
   */
  private const val DEBOUNCE_MS = 8_000L

  /**
   * Long enough for a redial to get through a connect and a handshake, short
   * enough that a subnet full of forged packets cannot hold the phone awake.
   */
  private const val WAKE_MS = 20_000L

  @Volatile
  private var thread: Thread? = null

  @Volatile
  private var socket: DatagramSocket? = null

  private var lock: WifiManager.MulticastLock? = null

  @Volatile
  private var lastAt = 0L

  /** Whether anything is listening, for the log line and for a test to ask. */
  val listening: Boolean
    get() = thread != null

  @Synchronized
  fun start(context: Context) {
    if (thread != null) return
    val app = context.applicationContext
    val bound = try {
      // `reuseAddress` before `bind`, so a service restarted into a process
      // whose old socket is still in TIME_WAIT comes back rather than giving
      // up on the feature until the next reboot.
      DatagramSocket(null).apply {
        reuseAddress = true
        broadcast = true
        bind(InetSocketAddress(PORT))
      }
    } catch (error: Exception) {
      // Another app on the port, or a policy that refuses the bind. The link
      // works exactly as it did before this existed, which is why this is a
      // warning and not a failure.
      Trace.warn("announce.bind.failed", "port" to PORT, "error" to error.javaClass.simpleName)
      return
    }
    socket = bound
    acquireLock(app)
    val worker = Thread({ listen(app, bound) }, "omarchy-announce")
    worker.isDaemon = true
    thread = worker
    worker.start()
    Trace.evt("announce.listening", "port" to PORT, "lock" to (lock?.isHeld == true))
  }

  @Synchronized
  fun stop() {
    val open = socket
    socket = null
    thread = null
    // Closing the socket is how the blocking receive is interrupted; the
    // thread falls out of its loop on the next line after that.
    try {
      open?.close()
    } catch (error: Exception) {
      /* already closed */
    }
    releaseLock()
    Trace.evt("announce.stopped")
  }

  private fun listen(context: Context, bound: DatagramSocket) {
    val buffer = ByteArray(MAX)
    while (true) {
      val packet = DatagramPacket(buffer, buffer.size)
      try {
        bound.receive(packet)
      } catch (error: Exception) {
        // The ordinary end of this thread: `stop()` closed the socket under
        // it. Anything else is worth a line, because a receive loop that has
        // quietly died looks exactly like a desktop that never announced.
        if (socket === bound) Trace.warn("announce.receive.failed", "error" to error.javaClass.simpleName)
        return
      }
      try {
        handle(context, packet)
      } catch (error: Exception) {
        Trace.warn("announce.handle.failed", "error" to error.javaClass.simpleName)
      }
    }
  }

  private fun handle(context: Context, packet: DatagramPacket) {
    val text = String(packet.data, packet.offset, packet.length, Charsets.UTF_8)
    val body = try {
      JSONObject(text)
    } catch (error: Exception) {
      // A UDP port on a busy subnet hears all sorts of things. Not news.
      Trace.detail("announce.ignored", "reason" to "unparseable")
      return
    }
    if (body.optString("app") != APP || body.optString("t") != KIND) {
      Trace.detail("announce.ignored", "reason" to "not-ours")
      return
    }
    val now = System.currentTimeMillis()
    if (now - lastAt < DEBOUNCE_MS) {
      Trace.detail("announce.ignored", "reason" to "debounced")
      return
    }
    lastAt = now

    val from = packet.address?.hostAddress
    // The address is logged, the key is not: it identifies a desktop, and this
    // log is read off a phone somebody else may be holding.
    Trace.evt("announce.heard", "from" to Trace.mark(from), "port" to body.optInt("port", 0))

    // The CPU is awake because a packet arrived, and stays awake exactly that
    // long unless somebody says otherwise. Redialling is a connect and a
    // handshake, so this says otherwise — for twenty seconds, on its own
    // timeout, whatever happens next.
    hold(context)

    // Nothing is written down for later, and nothing should be: an
    // announcement is only worth acting on in the seconds after it arrives,
    // and a phone whose runtime is gone has no link to bring back — the
    // service starting one is what would deliver the next announcement anyway.
    LinkActionReceiver.listener?.let { emit ->
      emit(
        "onDesktopAnnounce",
        mapOf(
          "app" to body.optString("app"),
          "t" to body.optString("t"),
          "name" to body.optString("name").ifBlank { null },
          "host" to body.optString("host").ifBlank { null },
          "port" to body.optInt("port", 0).takeIf { it > 0 },
          "publicKey" to body.optString("publicKey").ifBlank { null },
          "fingerprint" to body.optString("fingerprint").ifBlank { null },
          "protocol" to body.optInt("protocol", 0).takeIf { it > 0 },
          "version" to body.optString("version").ifBlank { null },
          "from" to from,
        ),
      )
    } ?: Trace.detail("announce.parked", "reason" to "no-runtime")
  }

  /** A short, self-releasing hold on the processor. Never an untimed one. */
  private fun hold(context: Context) {
    val power = context.getSystemService(PowerManager::class.java) ?: return
    try {
      val wake = power.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "omarchy:announce")
      wake.setReferenceCounted(false)
      wake.acquire(WAKE_MS)
    } catch (error: Exception) {
      // Without it the redial may not finish before the phone sleeps again,
      // which is the old behaviour rather than a new failure.
      Trace.warn("announce.wakelock.failed", "error" to error.javaClass.simpleName)
    }
  }

  private fun acquireLock(context: Context) {
    if (lock != null) return
    try {
      val wifi = context.getSystemService(WifiManager::class.java) ?: return
      val held = wifi.createMulticastLock("omarchy-announce")
      held.setReferenceCounted(false)
      held.acquire()
      lock = held
    } catch (error: Exception) {
      // Some devices deliver subnet broadcasts without it. The listener is
      // still worth running: it costs nothing and it works on the ones that do.
      Trace.warn("announce.lock.failed", "error" to error.javaClass.simpleName)
    }
  }

  private fun releaseLock() {
    val held = lock ?: return
    lock = null
    try {
      if (held.isHeld) held.release()
    } catch (error: Exception) {
      /* released under us, or never really held */
    }
  }
}

File: app/src/lib/announce.ts
/**
 * What the phone does about a desktop shouting "I am up" on the subnet.
 *
 * The desktop broadcasts one short burst when its daemon starts — see the
 * daemon's `lib/announce` for why it has to be a broadcast and not a message
 * addressed to this handset. The phone's own retry ladder tops out at two
 * minutes and then knocks on that rung forever, and on a sleeping phone even
 * that is a fiction: the timers are `setTimeout`s in a headless task holding
 * no wake lock, so the next attempt happens whenever something else wakes the
 * phone. The announcement is what makes the desktop's start-up be that
 * something.
 *
 * Everything here is the part that needs no socket and no Android: which
 * packets are worth acting on. It is deliberately a pure function of the
 * packet and what the phone already knows, because the whole risk of this
 * feature lives in that decision — a broadcast frame can be forged by anyone
 * on the network, and the answer to a forged one must be "nothing happened".
 *
 * Note what acting on it *is*: `reconnectNow(true)`, and nothing else. The
 * packet is never a reason to trust an address, a key or a certificate. The
 * phone dials the desktop it already pinned, over the endpoints it already
 * has, and `isOurDesktop` decides who is on the other end exactly as before.
 * The worst a stranger achieves by forging one is that the phone tries the
 * desktop it wanted to reach anyway.
 */

/** Where the phone listens. Mirrors `ANNOUNCE_PORT` in the daemon. */
export const ANNOUNCE_PORT = 8766

const ANNOUNCE_APP = 'omarchy-connect'
const ANNOUNCE_KIND = 'desktop-up'

/**
 * The packet as it arrives, plus `from` — the source address the native
 * receiver read off the datagram, which is not in the payload and cannot be
 * written by whoever composed it.
 */
export type DesktopAnnouncement = {
  app?: string | null
  t?: string | null
  protocol?: number | null
  version?: string | null
  name?: string | null
  host?: string | null
  port?: number | null
  publicKey?: string | null
  fingerprint?: string | null
  from?: string | null
}

/** What the phone knows about the desktop it is paired with, right now. */
export type LinkFacts = {
  /** The identity key pinned at pairing, as hex, or `null` when unpaired. */
  publicKey: string | null
  /** Whatever the socket is doing — `connected` means there is nothing to do. */
  status: string
}

/**
 * Whether this packet is reason enough to dial right now.
 *
 * Four ways to be ignored, and each of them is a case that happened to
 * somebody:
 *
 *   - **Not ours to read.** Anything on the port that is not this app's
 *     announcement, including the noise a busy subnet puts on any UDP port.
 *   - **Nobody to dial.** A phone with no pinned key has not paired, and there
 *     is no desktop for it to come back to.
 *   - **Somebody else's desktop.** Two Omarchy boxes on one network is a
 *     normal thing to have, and the flatmate's daemon restarting is not this
 *     phone's business. The key in the packet is a filter and not a
 *     credential: it saves a socket, it does not grant one.
 *   - **Already there.** The link is up; the announcement is stale news, and
 *     a redial would be dropping a working socket for a hope. This is the
 *     cautious half of the open question in the issue — the packet is heard
 *     only while the link is down.
 */
export function shouldRedial(packet: unknown, facts: LinkFacts): boolean {
  const announce = packet as DesktopAnnouncement | null
  if (!announce || typeof announce !== 'object') return false
  if (announce.app !== ANNOUNCE_APP || announce.t !== ANNOUNCE_KIND) return false
  if (!facts.publicKey) return false
  if (typeof announce.publicKey !== 'string' || announce.publicKey !== facts.publicKey) return false
  return facts.status !== 'connected'
}
